Exareme-Docker/src/exareme/exareme-master/src/main/java/madgik/exareme/master/gateway/async/handler/HttpAsyncMiningQueryHelper.java

```diff
@@ -0,0 +1,293 @@
+        }
+        if (!parameters.isEmpty()) {
+            log.debug("All of the parameters: " + parameters);
+            for (Map k : parameters) {
+                String name = (String) k.get("name");
+                String value = (String) k.get("value");
+                if (name == null || name.isEmpty() || value == null || value.isEmpty()) continue;
+
+                log.debug("Parameter in the json: ");
+                log.debug(name + " = " + value);
+
+                value = value.replaceAll("[^A-Za-z0-9,._~*+><=&|(){}:\\-\\\"\\[\\]]", "");    // ><=&| we no more need those for filtering
+                value = value.replaceAll("\\s+", "");
+
+                log.debug("Parameter after format: ");
+                log.debug(name + " = " + value);
+
+                inputContent.put(name, value);
+            }
+            return inputContent;
+        }
+        return null;
+    }
+
+
+    public static String searchConsul(String query) throws IOException {
+        String result = null;
+        CloseableHttpClient httpclient = HttpClients.createDefault();
+        String consulURL = System.getenv("CONSULURL");
+        if (consulURL == null) throw new IOException("Consul url not set");
+        if (!consulURL.startsWith("http://")) {
+            consulURL = "http://" + consulURL;
+        }
+
+        HttpGet httpGet;
+        httpGet = new HttpGet(consulURL + "/v1/kv/" + query);
+        log.debug("Running: " + httpGet.getURI());
+        CloseableHttpResponse response = null;
+        if (httpGet.toString().contains(System.getenv("EXAREME_MASTER_PATH") + "/") || httpGet.toString().contains(System.getenv("DATA") + "/")) {    //if we can not contact : http://exareme-keystore:8500/v1/kv/master* or http://exareme-keystore:8500/v1/kv/datasets*
+            try {   //then throw exception
+                response = httpclient.execute(httpGet);
+            } catch (Exception e) {
+                response.close();
+            }
+            result = EntityUtils.toString(response.getEntity());
+        }
+        if (httpGet.toString().contains(System.getenv("EXAREME_ACTIVE_WORKERS_PATH") + "/")) {    //if we can not contact : http://exareme-keystore:8500/v1/kv/active_workers*
+            //then maybe there are no workers running
+            try {
+                response = httpclient.execute(httpGet);
+                if (response.getStatusLine().getStatusCode() != 200) {
+                    if (httpGet.toString().contains("?keys"))
+                        log.debug("No workers running. Continue with master");
+                } else {
+                    result = EntityUtils.toString(response.getEntity());
+                }
+            }
+            catch (Exception e){
+                response.close();
+            }
+        }
+        return result;
+    }
+
+    public static List<String> getInputAlgo(HashMap<String, String> inputContent) throws IOException, DatasetsException, PathologyException {
```

Comment:
You can refactor this function to return a List<ContainerProxy> so on the pingContainer you won't have to search again for the container.